packages/nimble-components/src/table/specs/data-hierarchy-hld.md

```diff
@@ -0,0 +1,121 @@
+
+I see no reason to explicitly disable this behavior. One behavior we will need to ensure, however, is that the count value we display in the group row, is _only_ the number of immediate children in the group row. It would be odd for the number to change just because a row was expanded (such as in the lazy loading case).
+
+#### Managing expanded state
+
+Currently, the `Table` defaults to expanding all rows by setting its Tanstack expanded state to the singleton `true`. This means that all group rows will be expanded by default. The Nimble `Table` also provides an implementation of `getIsRowExpanded` which is consumed by `TanStack` that will ultimately result in the per-row state of whether the row is expanded. It is in this implementation, where even if the Tanstack state is set to `true`, we can mark all parent rows, by default, as collapsed, which is required for the lazy-loading scenario.
+
+This is achievable because the Nimble `Table` currently tracks when particular rows are collapsed by their row id, which both group and parent rows have. When the Tanstack state has the singleton value of `true` we know we are in a default expand/collapsed state (meaning the user hasn't interactively changed anything), as otherwise it will be a set of id values matched with a boolean state. This will allow us to implement a behavior in `getIsRowExpanded` that will denote group rows as expanded, but parent rows as collapsed, specifically when the TanStack expanded state is set to `true`.
+
+Once _any_ row has been expanded or collapsed, we must update the Nimble `Table` state where we track collapsed rows with _all_ rows that are currently collapsed. This will be a one-time cost. Prototyping suggests that the performance penalty isn't that noticeable.
+
+### Showing a progress indicator for lazy loading
+
+After a user clicks on a parent row to expand it, and that parent does not have children yet, there could be a noticeable delay between when the user clicked the expand button, and when the children show up in the table. This could happen for a variety of reasons, but commonly might occur due to a slow network request.
+
+It would be helpful for the `Table` to visually represent the state that it is currently waiting on data to be loaded from the client. It can accomplish this fairly easily using existing state with no additional APIs or new components.
+
+Prototype visual:
+
+![Lazy Loading Spinner](./spec-images/LazyLoadingSpinner.gif)
+
+This ultimately may put the burden on the client to ensure that the `Table` is updated as needed to get rid of any displayed progress indicator, including in scenarios where the expansion of a parent row failed to load any children (possibly due to some client-side error). The `Table` may only guarantee that the progress indicator is shown when a parent row is expanded and it currently has no children, and that it will be removed once children are present.
```

Comment:
I think it would be good to harden the proposal here (or maybe just the language associated with it). Specifically, I think we should state when the table _will_ (not _may_) put a loading affordance for the row. Then, maybe you could talk about how in the expected workflow the loading spinner would be removed automatically and how a client could remove it in some unexpected cases (like a failed network request or lazy-loading the children determined there aren't actually children like was initially thought).

Sorry if this seems nit-picky, but it's a little unclear to me right now whether this is actually the proposal or just one possible solution to show that we have options.

Reply:
I've made the wording a bit stronger and added some workflows.